File: Programmers/java/LEVEL_3_부대복귀.java
import java.util.*;

class Solution {
    private void bfs(Map<Integer, List<Integer>> map, boolean[] visit, int destination, Set<Integer> sources, int[] tempAnswer) {
        Queue<int[]> q = new LinkedList<>(); //현재 위치, 현재까지의 이동 거리
        q.add(new int[] {destination, 0});
        visit[destination] = true;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int cur = current[0];
            int cnt = current[1];

            if (sources.contains(cur)) {
                tempAnswer[cur] = cnt;
                sources.remove(cur);
            }

            List<Integer> area = map.get(cur);
            for (int i = 0; i < area.size(); i++) {
                if (visit[area.get(i)] == false) {
                    q.add(new int[] {area.get(i), cnt + 1});
                    visit[area.get(i)] = true;
                }
            }
        }
    }

    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        Map<Integer, List<Integer>> map = new HashMap<>();
        boolean[] visit = new boolean[n + 1];
        Set<Integer> sourcesSet = new HashSet<>();
        int[] tempAnswer = new int[n + 1];

        Arrays.fill(tempAnswer, -1);

        for (int source : sources) {
            sourcesSet.add(source);
        }


        for (int[] road : roads) {
            if (!map.containsKey(road[0])) {
                map.put(road[0], new ArrayList<>());
            }
            map.get(road[0]).add(road[1]);

            if (!map.containsKey(road[1])) {
                map.put(road[1], new ArrayList<>());
            }
            map.get(road[1]).add(road[0]);
        }

        bfs(map, visit, destination, sourcesSet, tempAnswer);

        int idx = 0;
        for (int i = 0; i < sources.length; i++) {
            answer[idx++] = tempAnswer[sources[i]];
        }

        return answer;
    }
}

File: Programmers/cpp/LEVEL_3_부대복귀.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <queue>
#include <set>

using namespace std;

void bfs(unordered_map<int, vector<int>>& um, vector<bool>& visit, int destination, set<int>& sources, vector<int>& temp_answer) {
	queue<pair<int, int>> q; // 현재 위치, 현재까지의 이동 거리
	q.push({ destination, 0 });
	visit[destination] = true;

	while (!q.empty()) {
		int cur = q.front().first;
		int cnt = q.front().second;
		q.pop();

		if (sources.find(cur) != sources.end()) {
			temp_answer[cur] = cnt;
			sources.erase(cur);
		}

		vector<int> area = um[cur];
		for (int i = 0; i < area.size(); i++) {
			if (visit[area[i]] == false) {
				q.push({ area[i], cnt + 1 });
				visit[area[i]] = true;
			}
		}
	}
}

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
	vector<int> answer;
	unordered_map<int, vector<int>> um;
	vector<bool> visit(n + 1, false);
	set<int> source_set(sources.begin(), sources.end());
	vector<int> temp_answer(n + 1, -1);

	for (vector<int> road : roads) {
		um[road[0]].push_back(road[1]);
		um[road[1]].push_back(road[0]);
	}

	bfs(um, visit, destination, source_set, temp_answer);

	for (int i = 0; i < sources.size(); i++) {
		answer.push_back(temp_answer[sources[i]]);
	}

	return answer;
}